src/Compilers/CSharp/Portable/Binder/DecisionDagBuilder.cs

```diff
@@ -1762,31 +1915,137 @@ public sealed class One : Tests
+                    Debug.Assert(sideeffect is True or One(BoundDagAssignmentEvaluation));
+
+                    // Given:
+                    //
+                    //  - "test" as a test that has already occurred,
+                    //  - "other" as a subsequent test,
+                    //  - S as a possible side-effect (expected to always evaluate to True),
+                    //  - and P as a pre-condition under which we need to evaluate S,
+                    //
+                    // we proceed as follows:
+                    //
+                    //  - If "test" being true proves "other" to be also true, we rewrite "other" as ((P && S) || other),
+                    //    because we have determined that on this branch, "other" would always succeed if the pre-condition is met.
+                    //    Note: If there is no pre-condition, i.e. P is True, the above will be reduced to True which means "other" is insignificant.
+                    //
+                    //  - If "test" being true proves "other" to be false, we rewrite "other" as (!(P && S) && other),
+                    //    because we have determined that on this branch, "other" would never succeed if the pre-condition is met.
+                    //    Note: If there is no pre-condition, i.e. P is True, the above will be reduced to False which means "other" is impossible.
+                    //
+                    //  - Otherwise, we rewrite "other" as ((!P || S) && other) to preserve the side-effect if the pre-condition is met,
+                    //    because we have determined that there were no logical implications from one to the other on this branch.
+                    //    Note: If there is no pre-condition, i.e. P is True, "other" is not rewritten which means the two are considered independent.
+                    //
+                    whenTrue = rewrite(trueDecisionImpliesTrueOther, trueDecisionPermitsTrueOther, precondition, sideeffect, this);
+
+                    // Similarly for the opposite branch when "test" is false.
+                    whenFalse = rewrite(falseDecisionImpliesTrueOther, falseDecisionPermitsTrueOther, precondition, sideeffect, this);
+
+                    static Tests rewrite(bool decisionImpliesTrueOther, bool decisionPermitsTrueOther, Tests precondition, Tests sideeffect, Tests other)
+                    {
+                        return decisionImpliesTrueOther
+                            ? OrSequence.Create(AndSequence.Create(precondition, sideeffect), other)
+                            : !decisionPermitsTrueOther
+                                ? AndSequence.Create(Not.Create(AndSequence.Create(precondition, sideeffect)), other)
+                                : AndSequence.Create(OrSequence.Create(Not.Create(precondition), sideeffect), other);
+                    }
```

Comment:
We had a design review with the dev team on this feature. The goal was mostly education (refresh the team based on what Fred and I now understand about DAG design and explain the changes from this PR) and also allow feedback.
@AlekseyTs raised an interesting idea which I'd like to bounce with you. This is not meant to block this PR, but in case it could spark your imagination with a potentially simpler design.

Instead of injecting preconditions and sideeffects, what if this `Filter` method had the following structure:
// step 1: if the length is known to have a single value (**), then rewrite the current test in light of that knowledge
// step 2: existing logic `CheckConsistentDecision` followed by simple ternaries (as it existed prior to this PR)

In step 1: if in the current state we know that Length==1 (by looking at remaining values), and the current test/evaluation is regarding an indexer access, then we replace the current test. For example `t3 == 42` would be rewritten as `t2 == 42`. Basically, from this point on in the DAG we'd be dealing with normalized temps (no longer dealing with `t3`, and dealing with `t2` instead).

We'd need to think more about how to deal with remaining values in that design. Conceptually a similar approach might work (if the length is known to have a specific value in the current state, then remaining values of certain temps can be merged).

(**) in writing this summary, I realized that identifying states where length gets narrowed to a single value isn't easy. If the selected test is `Length == k` then this is easy. But the selected test could be `Length >= k` which could narrow the set of remaining values to only `k`. So we'd have to compute remaining values first to see if `Length` was narrowed to a single value...

Reply:
>rewrite the current test in light of that knowledge

>we'd have to compute remaining values first

Since we've been through many MANY flaws in the current design already (thanks for that btw!) I think we can easily poke holes in alternative approaches. IIRC we did consider both of these (which are interdependent) but I was convinced it's going to get too subtle. There's just many questions to answer before we even get close to a rough sketch. Here's some examples:

- What happens  if `RemainingTests` is not linear? A precomputed length value may not be always the same for all indexer evaluations e.g. when we have `or` combinators which will cause the dag to split. That would require a lot of plumbing to get right (remember those could come from an explicit Length test in a property pattern so there's infinite possibilities).

- Rewriting tests would have the same problem, plus we may still need the original input for subsequent tests so we will need to keep track of both somewhere. When you say "from this point on in the dag" that means we need to adjust DagState which has it's own way of handling we didn't need to touch in this PR. 

- At a given state, what temps need to be merged exactly? It's possible that the rewrite depends on multiple length values and we can't depend on a direct value test for the reasons mentioned above.

The current approach is formulated inside the decision dag itself, rather than a pre-processing step + maintaining a bunch of additional states to cover all corner cases. So I believe this has made as simple as possible, if we go any further there's a high chance that we could miss something.

Reply:
Also, we actually need the additional length tests to be emitted as demonstrated in speclet:

>- For instance, `[_, >0, ..] or [.., <=0, _]` becomes `length >= 2 && ([1] > 0 || length == 3 || [^2] <= 0)` where the length value of 3 implies the other test.
>- Conversely, `[_, >0, ..] and [.., <=0, _]` becomes `length >= 2 && [1] > 0 && length != 3 && [^2] <= 0` where the length value of 3 disallows the other test.